sympy/solvers/solveset.py

```diff
@@ -1723,6 +1733,27 @@ def linear_eq_to_matrix(equations, *symbols):
         # append constant term (term free from symbols)
         coeff_list.append(-f.as_coeff_add(*symbols)[0])
 
+        if options['check_linearity']:
```

Comment:
This reconstruction is not robust enough. Consider `eq = x + x*(y + 1)` which is linear in x but has a coefficient of x of `y + 2`; the expression that this code would test is `eq - x*(y + 2)` but that is not zero. There are some helpers like `degree` or `solve_linear` that are more careful about such equations.

Reply:
I have implemented solve_linear(). Now it works correctly for the equation said. However, there is a bug in coeff method. `(x + x*(y + 1)).coeff(y)` returns 0, but it should be x. Can I create an issue for this?

Reply:
> issue

No. coeff is fast and dumb by design, as stated in docstring "The matching is exact:" Your option is to first use expand on the equations -- that is the only way to know what the coefficients of the symbols are. We have not yet implemented a method of giving the exact order when there are power of monomials as in `x + (x + 1)**3`. This is perhaps why expansion is used by Poly at instantiation. If someone is working with a linear system, then the expansion step should be quick and assure you have accurate results.